cmd/bad-key-revoker/main.go

```diff
@@ -0,0 +1,360 @@
+	"context"
+	"crypto/x509"
+	"flag"
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	netmail "net/mail"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/letsencrypt/boulder/cmd"
+	"github.com/letsencrypt/boulder/core"
+	corepb "github.com/letsencrypt/boulder/core/proto"
+	"github.com/letsencrypt/boulder/db"
+	bgrpc "github.com/letsencrypt/boulder/grpc"
+	"github.com/letsencrypt/boulder/mail"
+	rapb "github.com/letsencrypt/boulder/ra/proto"
+	"github.com/letsencrypt/boulder/revocation"
+	"github.com/letsencrypt/boulder/sa"
+
+	"google.golang.org/grpc"
+)
+
+type revoker interface {
+	AdministrativelyRevokeCertificate(ctx context.Context, in *rapb.AdministrativelyRevokeCertificateRequest, opts ...grpc.CallOption) (*corepb.Empty, error)
+}
+
+type badKeyRevoker struct {
+	dbMap                 *db.WrappedMap
+	uncheckedBatchSize    int
+	certificatesBatchSize int
+	raClient              revoker
+	mailer                mail.Mailer
+}
+
+type unchecked struct {
+	ID        int
+	KeyHash   []byte
+	RevokedBy int64
+}
+
+func (bkr *badKeyRevoker) selectUncheckedRows() ([]unchecked, error) {
+	var uncheckedRows []unchecked
+	initialID := 0
+	for {
+		var batch []unchecked
+		_, err := bkr.dbMap.Select(
+			&batch,
+			"SELECT id, keyHash, revokedBy FROM blockedKeys WHERE extantCertificatesChecked = false AND id > ? ORDER BY id LIMIT ?",
+			initialID,
+			bkr.uncheckedBatchSize,
+		)
+		if err != nil {
+			if db.IsNoRows(err) {
+				return uncheckedRows, nil
+			}
+			return nil, err
+		}
+		uncheckedRows = append(uncheckedRows, batch...)
+		if len(batch) != bkr.uncheckedBatchSize {
+			return uncheckedRows, nil
+		}
+		initialID = batch[len(batch)-1].ID
+	}
+}
+
+type unrevoked struct {
+	Serial         string
+	DER            []byte
+	RegistrationID int64
+	RevokedBy      int64
+}
+
+var revokerName = "bad-key-revoker"
+
+func (bkr *badKeyRevoker) findUnrevoked(row unchecked) ([]unrevoked, error) {
+	var unrevokedCerts []unrevoked
+	initialID := 0
+	for {
+		var batch []struct {
+			ID         int
+			CertSerial string
+		}
+		_, err := bkr.dbMap.Select(
+			&batch,
+			"SELECT id, certserial FROM keyHashToSerial WHERE keyHash = ? and id > ? ORDER BY id LIMIT ?",
+			row.KeyHash,
+			initialID,
+			bkr.certificatesBatchSize,
+		)
+		if err != nil {
+			if db.IsNoRows(err) {
+				return unrevokedCerts, nil
+			}
+			return nil, err
+		}
+		for _, serial := range batch {
+			var unrevokedCert unrevoked
+			err = bkr.dbMap.SelectOne(
+				&unrevokedCert,
+				`SELECT cs.serial, c.registrationID, c.der FROM certificateStatus AS cs
+				JOIN certificates AS c
+				ON cs.serial = c.serial
+				WHERE cs.serial = ? AND cs.isExpired = false AND cs.status != ?`,
+				serial.CertSerial,
+				string(core.StatusRevoked),
+			)
+			if err != nil {
+				if db.IsNoRows(err) {
+					continue
+				}
+				return nil, err
+			}
+			unrevokedCert.RevokedBy = row.RevokedBy
+			unrevokedCerts = append(unrevokedCerts, unrevokedCert)
+		}
+		if len(batch) != bkr.certificatesBatchSize {
```

Comment:
Here we have two possible ways to exit the loop: (1) By getting an empty response from the select, or (2) by getting a response from the select that's less than the requested size. Can we reduce this to just the (1) case? If we get a short select, the next select should return no rows.